core/src/test/java/kafka/testkit/KafkaClusterTestKit.java

```diff
@@ -275,15 +275,15 @@ metaProperties, config, new MetadataRecordSerde(), metadataPartition, KafkaRaftS
                     executorService.shutdownNow();
                     executorService.awaitTermination(5, TimeUnit.MINUTES);
                 }
-                for (ControllerServer controller : controllers.values()) {
-                    controller.shutdown();
-                }
                 for (BrokerServer brokerServer : brokers.values()) {
                     brokerServer.shutdown();
                 }
                 for (KafkaRaftManager<ApiMessageAndVersion> raftManager : raftManagers.values()) {
                     raftManager.shutdown();
                 }
+                for (ControllerServer controller : controllers.values()) {
+                    controller.shutdown();
+                }
```

Comment:
Is it right that shutdown and startup are in the same order? That is, we don't shutdown in the reverse order of startup?

Reply:
If this is correct, we should probably add a comment explaining it as it's not obvious why it would work that way.

Reply:
I was trying to make the order consistent with `KafkaRaftServer`. The circular dependence between `RaftManager` and `ControllerServer` makes this a bit awkward in any case. `RaftManager` needs the `SocketServer` from `ControllerServer`. On the other hand, `ControllerServer` needs to register a listener with `RaftManager`. Perhaps we need to pull `SocketServer` up to a higher level? But that would be a major effort.

Reply:
I added a few comments to clarify. Really I think `KafkaClusterTestKit` should be using `KafkaRaftServer` directly instead of building the components directly. I've wanted to do this refactor for a while, but have not had time.